incentives.py

```diff
@@ -0,0 +1,311 @@
+        # pass in model, decision nodes and utility nodes of agent being considered
+        try:
+            assert isinstance(model, BayesianModel)
+        except AssertionError:
+            raise AssertionError("Control instantiation's first argument must be a BayesianModel")
+
+        self.dag = model
+        self.graph = self.dag.to_directed()
+        self.decision_nodes = decision_nodes
+        self.utility_nodes = utility_nodes
+        self.all_nodes = self.dag.nodes
+
+        try:
+            assert set(self.decision_nodes).issubset(set(self.all_nodes))
+            assert set(self.utility_nodes).issubset(set(self.all_nodes))
+        except AssertionError:
+            raise AssertionError("Some or all of the Decision and Utility nodes are not present on the graph")
+
+    def con_node(self, node):
+        # returns True if a node faces a control incentive
+        trimmed_dag = TrimGraph(self.dag, self.decision_nodes, self.utility_nodes).trim()
+        if node != self.decision_nodes[0]:
+            if node == self.utility_nodes[0] or self.utility_nodes[0] in nx.descendants(trimmed_dag, node):
+                return True
+            else:
+                return False
+
+    def indir_con_node(self, node): #!!! need to update to this to reflect new def in UndAgInc paper (backdoor)
+        # returns True if a node faces an intervention incentive for control
+        trimmed_dag = TrimGraph(self.dag, self.decision_nodes, self.utility_nodes).trim()
+        if self.int_node(node):
+            x_u_paths = Paths(trimmed_dag).find_all_dir_path(node, self.utility_nodes[0])
+            if any(self.decision_nodes[0] in paths for paths in x_u_paths):
+                return True
+            else:
+                return False
+
+    def dir_con_node(self, node):
+        # returns True if a node faces a direct control incentive
+        trimmed_dag = TrimGraph(self.dag, self.decision_nodes, self.utility_nodes).trim()
+        if self.int_node(node):
+            x_u_paths = Paths(trimmed_dag).find_all_dir_path(node, self.utility_nodes[0])
+            for path in x_u_paths:
+                if self.decision_nodes[0] not in path:
+                    return True
+            return False
+
+
+    def all_cont_inc_nodes(self):
+        con_list = []
+        for x in list(self.dag.nodes):
+            if self.con_node(x):
+                con_list.append(x)
+        return con_list
+
+    def all_dir_con_inc_nodes(self):
+        dir_con_list = []
+        for x in list(self.dag.nodes):
+            if self.dir_con_node(x):
+                dir_con_list.append(x)
+        return dir_con_list
+
+    def all_indir_con_inc_nodes(self):
+        indir_con_list = []
+        for x in list(self.dag.nodes):
+            if self.indir_con_node(x):
+                indir_con_list.append(x)
+        return indir_con_list
+
+##
+class Paths:
+    # class for finding paths that exist in the MAID
+    def __init__(self, model):
```

Comment:
I don't think it makes sense for this to be a class that takes a model in. It makes more sense to me that it should be a MixIn that CID and MACID inherit from, because it represents things that a CID and MACID can both do.